src/main/java/application/MenuController.java

```diff
@@ -87,34 +120,42 @@ private SessionBean getNextMenu(MenuSession menuSession, boolean redrawing) thro
             return menuSession.startFormEntry(sessionRepo);
         }
         else{
-            MenuResponseBean menuResponseBean = new MenuResponseBean();
-            menuResponseBean.setSessionId(menuSession.getSessionId());
+            MenuSessionBean menuResponseBean = new MenuSessionBean();
+
             // We're looking at a module or form menu
             if(nextScreen instanceof MenuScreen){
-                MenuScreen menuScreen = (MenuScreen) nextScreen;
-                menuResponseBean.setMenuType(Constants.MENU_MODULE);
-                menuResponseBean.setOptions(getMenuRows(menuScreen));
+                menuResponseBean = generateMenuScreen((MenuScreen) nextScreen);
             }
             // We're looking at a case list or detail screen (probably)
             else if (nextScreen instanceof EntityScreen) {
-                EntityScreen entityScreen = (EntityScreen) nextScreen;
-                menuResponseBean.setMenuType(Constants.MENU_ENTITY);
-                menuResponseBean.setOptions(getMenuRows(entityScreen.getCurrentScreen()));
+                menuResponseBean = generateEntityScreen((EntityScreen) nextScreen);
+            } else{
+                throw new Exception("What screen are we on? " + nextScreen);
             }
+            menuResponseBean.setSessionId(menuSession.getSessionId());
             return menuResponseBean;
         }
     }
 
-    private HashMap<Integer, String> getMenuRows(OptionsScreen nextScreen){
-        String[] rows = nextScreen.getOptions();
-        HashMap<Integer, String> optionsStrings = new HashMap<Integer, String>();
-        for(int i=0; i <rows.length; i++){
-            optionsStrings.put(i, rows[i]);
-        }
-        return optionsStrings;
+    private CommandListResponseBean generateMenuScreen(MenuScreen nextScreen){
+        return new CommandListResponseBean(nextScreen);
     }
 
+    private EntityListResponse generateEntityScreen(EntityScreen nextScreen){
+        return new EntityListResponse(nextScreen);
+    }
     private MenuSession getMenuSession(String sessionId) throws Exception {
         return new MenuSession(menuRepo.find(sessionId), restoreService, installService);
     }
+
+    @ExceptionHandler(Exception.class)
+    public String handleError(HttpServletRequest req, Exception exception) {
+        log.error("Request: " + req.getRequestURL() + " raised " + exception);
+        exception.printStackTrace();
```

Comment:
this still feels weird to me, but i don't really know how to better handle errors in java